mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeans.scala

```diff
@@ -290,29 +311,30 @@ class KMeans private (
       val bcActiveCenters = sc.broadcast(activeCenters)
 
       // Find the sum and count of points mapping to each center
-      val totalContribs = data.mapPartitions { points =>
-        val thisActiveCenters = bcActiveCenters.value
-        val runs = thisActiveCenters.length
-        val k = thisActiveCenters(0).length
-        val dims = thisActiveCenters(0)(0).vector.size
-
-        val sums = Array.fill(runs, k)(Vectors.zeros(dims))
-        val counts = Array.fill(runs, k)(0L)
-
-        points.foreach { point =>
-          (0 until runs).foreach { i =>
-            val (bestCenter, cost) = KMeans.findClosest(thisActiveCenters(i), point)
-            costAccums(i) += cost
-            val sum = sums(i)(bestCenter)
-            axpy(1.0, point.vector, sum)
-            counts(i)(bestCenter) += 1
+      val totalContribs = data.sample(false, miniBatchFraction, 42 + iteration)
```

Comment:
Thanks for your reply. RJ Nowling also had suggested this point.
I will update this PR to expose `setMiniBatchFraction` in ml's KMeans when at the right time. And I will make sure when fraction is set 1, there is no overhead.
By the way, MiniBatch EM is also available in GMM: http://www.cs.ubc.ca/~murphyk/Teaching/CS340-Fall06/reading/mixtureModels.pdf
![wechatimg29](https://cloud.githubusercontent.com/assets/7322292/20045694/90088be8-a4de-11e6-8490-cdfcf9e2b505.jpeg)